src/backend/distributed/commands/multi_copy.c

```diff
@@ -2762,19 +2762,87 @@ ProcessCopyStmt(CopyStmt *copyStatement, char *completionTag, const char *queryS
 				 * COPY table TO PROGRAM / file is handled by wrapping the table
 				 * in a SELECT * FROM table and going through the result COPY logic.
 				 */
-				ColumnRef *allColumns = makeNode(ColumnRef);
 				SelectStmt *selectStmt = makeNode(SelectStmt);
-				ResTarget *selectTarget = makeNode(ResTarget);
 
+#if PG_VERSION_NUM >= PG_VERSION_12
+				List *targetList = NIL;
+				Relation distributedRelation = heap_openrv(copyStatement->relation,
+														   AccessShareLock);
+				TupleDesc tupleDescriptor = RelationGetDescr(distributedRelation);
+				bool hasGeneratedColumns = false;
+
+				for (int i = 0; i < tupleDescriptor->natts; i++)
+				{
+					Form_pg_attribute attr = &tupleDescriptor->attrs[i];
+					if (attr->attisdropped)
+					{
+						continue;
+					}
+
+					if (attr->attgenerated)
+					{
+						hasGeneratedColumns = true;
+						break;
+					}
+				}
+
+				if (hasGeneratedColumns)
+				{
+					for (int i = 0; i < tupleDescriptor->natts; i++)
+					{
+						Form_pg_attribute attr = &tupleDescriptor->attrs[i];
+
+						if (attr->attisdropped || attr->attgenerated)
+						{
+							continue;
+						}
+
+						ColumnRef *column = makeNode(ColumnRef);
+						column->fields = list_make1(makeString(pstrdup(
+																   attr->attname.data)));
+						column->location = -1;
+
+						ResTarget *selectTarget = makeNode(ResTarget);
+						selectTarget->name = NULL;
+						selectTarget->indirection = NIL;
+						selectTarget->val = (Node *) column;
+						selectTarget->location = -1;
+
+						targetList = lappend(targetList, selectTarget);
+					}
+
+					selectStmt->targetList = targetList;
+				}
+				else
+				{
+					ColumnRef *allColumns = makeNode(ColumnRef);
+					allColumns->fields = list_make1(makeNode(A_Star));
+					allColumns->location = -1;
+
+					ResTarget *selectTarget = makeNode(ResTarget);
+					selectTarget->name = NULL;
+					selectTarget->indirection = NIL;
+					selectTarget->val = (Node *) allColumns;
+					selectTarget->location = -1;
+
+					selectStmt->targetList = list_make1(selectTarget);
+				}
```

Comment:
Is there a performance benefit of having this special codepath that selects all columns, instead of _always_ creating the select targetlist by skipping deleted and generated columns?

If there is no performance gain at runtime I think it makes a lot of sense to remove this block. This would also remove the first iteration through the attributes to check if there is a generated column.

Lastly it could isolate the pg12 specific code to the `attr->attgenerated` check.
Now there is a lot of different codepaths to understand depending on pg11/pg12 and having generated columns vs not having generated columns.

Having 1 flow of selecting every column but skip deleted columns, and for pg12 skip generated columns as well will be easier to maintain and understand IMHO.